### PR TITLE
Normalise string prefix order

### LINKED ISSUE #2171: Normalize raw f-strings prefix
**Is your feature request related to a problem? Please describe.**

Black should normalize the prefix used for raw f-strings (https://www.python.org/dev/peps/pep-0498/#raw-f-strings). Both `rf` and `fr` work, but one should be preferred.

Currently, the following code is not changed:

```
    f = rf"hello {world}"
    f = fr"hello {world}"
```

**Describe the solution you'd like**

Pick one and normalize. I don't really care which. `rf` matches the name "raw f-string", but `fr` matches the example in PEP-498 and Jupyterlab syntax highlighting seems to prefer it.

**Describe alternatives you've considered**

Doing nothing leaves open the opportunity for inconsistency and disagreement.

**Additional context**

None

### PR DESCRIPTION
**Pull request by felix-hilden**
Closes #2171: this PR makes _Black_ normalise string prefix order. Since no other maintainer but Jelle responded to the last poll, I went ahead and put <strike>"r" first since it was also the winner of the first poll. But that's easy to change if you'd like.</strike> (<strike>"r" last now, see issue and below</strike> ("r" first now again lol))

- [x] Tests and changed existing tests and source to comply
- [x] Changelog and style documentation
- [x] 100% money back guarantee

---
Also, don't worry: I wrote "normalize" to the source for consistency. And for freedom! 🦅

### CODE REVIEW

File: tests/data/string_prefixes.py
Comment (cooperlees) [inline tests/data/string_prefixes.py:15]: NIT: I feel we should just go to python3 here
```suggestion
#!/usr/bin/env python3
```

File: src/black/strings.py
Comment (cooperlees) [inline src/black/strings.py:141]: Just double checking - There is never a chance of > 2 here right?

- Also, lets set a const for 2 to explain it's value

Feel free to pick a better name etc.
```suggestion
    TWO_CHAR_PREFIX_LEN = 2
    if len(new_prefix) == TWO_CHAR_PREFIX_LEN and "r" != new_prefix[0].lower():
```

**Review by cooperlees — APPROVED** [review]
LGTM. Just some unimportant suggestions + NITs.

File: tests/data/string_prefixes.py
Comment (felix-hilden) [reply tests/data/string_prefixes.py]: Consider it done!

File: src/black/strings.py
Comment (felix-hilden) [reply src/black/strings.py:141]: Yeah about that, at least with Python 3 it's max 2. Never used Python 2 tho, but trying on an online interpreter, "rb" is invalid and "f" doesn't exist. "ru" and "ub" are also invalid. So we should be ok.

I'll fight you a bit with adding the constant, if I may. Having "two char prefix len" be 2 is both a bit too obvious and not explaining enough in my opinion. Would `len(pf) > 1` or maybe a comment saying max 2 is guaranteed be better? I won't die on this hill though 😄 so feel free to just put your foot down.

Comment (cooperlees) [reply src/black/strings.py:141]: I also don't care enough to fight this. I've just always found it easier to come back to code with consts explaining a naked integer years / even weeks later. Happy with this if other maintainers are.

Comment (felix-hilden) [reply src/black/strings.py:142]: Added a comment there explaining it a bit more, that is definitely appropriate and enables a more verbose explanation. Will add the const if others also agree!

Comment (Numerlor) [reply src/black/strings.py:142]: 2 Seems to be the max according to the lexical analysis page on both versions.
I think a comment would be best here, as a var name won't imo explain why it's two without being overly verbose.

While not directly related, is there any reason for the yoda condition? `... new_prefix[0].lower() != "r"` reads better to me

Comment (felix-hilden) [reply src/black/strings.py:142]: I guess "r is not the first prefix char" seemed more intuitive for me when writing it, but it could be "the first char is not r" just as well!

File: tests/data/python2.py
Comment (JelleZijlstra) [inline tests/data/python2.py:12]: Why are the `ur` removed here?

Comment (felix-hilden) [reply tests/data/python2.py:12]: Oh wow. "ru" is invalid but "ur" isn't? 😅 Good catch. Let me go through the others once more.

Comment (felix-hilden) [reply tests/data/python2.py:12]: "br" is valid and "rb" invalid. Well. This is a bummer. They go exactly against the rule. Should we go with "r last" after all?

Comment (JelleZijlstra) [reply tests/data/python2.py:12]: Oh you're right. And in Python 3 they're both invalid.

Comment (felix-hilden) [reply tests/data/python2.py:12]: Yeah "u" with anything else seems to be invalid, but "br" naturally works.

Comment (JelleZijlstra) [reply tests/data/python2.py:12]: Sorry that comment was in reference to `ru` not `rb`.

In summary:
- Python 3 allows `rb`, `br`, but not `ur` or `ru`
- Python 2 allows `br` and `ur` but not `rb` or `ru`

Comment (felix-hilden) [reply tests/data/python2.py:12]: I commented on the original issue and marked this PR as a draft. We ought to discuss it through before merging anything.

File: src/black/strings.py
Comment (jamesmyatt) [reply src/black/strings.py:142]: What about some logic along the lines of "if there's an r in the prefix, then move it to the front" which will also work for future multi-character prefixes? You could also put the rest of the prefix in alphabetical order at the same time for future consistency. Might be gold-plating though.

Comment (felix-hilden) [reply src/black/strings.py:142]: Yeah, reversing is essentially a shorthand in that case. I don't expect many more prefixes to be added so I just left it at that. Are there any on the horizon? If not, I'd say it's just making it too fancy, although strictly correct 😄

**Comment by felix-hilden** [comment]
In the absense of comments from maintainers on the linked issue I went ahead and changed the rule to "r" last to take Python 2 into account more easily. As mentioned in the issue, since there were no strong opinions either way I think this is a reasonable switch.

File: src/black/strings.py
Comment (felix-hilden) [inline src/black/strings.py:143]: Sadly `sorted` isn't quite enough here because of `ur` :/

File: src/black/strings.py
Comment (felix-hilden) [inline src/black/strings.py:142]: And technically the first condition is redundant, but I like it because it clarifies the meaning.

**Comment by felix-hilden** [comment]
A friendly bump! My reviews above are only remarks, so I think this should be ready for review.

File: tests/data/string_prefixes.py
Comment (JelleZijlstra) [inline tests/data/string_prefixes.py:3]: Any reason to change the input section here?

Comment (felix-hilden) [reply tests/data/string_prefixes.py:3]: Just for reducing a redundant test, since all the prefixes are tested below! I thought about removing it altogether, but seems fine to test f-string substitution there at least. Would you like to keep it?

File: src/black/strings.py
Comment (Numerlor) [reply src/black/strings.py:143]: Could use a key like this with `sorted`, I think it's nicer than the reverse but I'm not sure if it reads better overall.
```py
new_prefix = "".join(sorted(new_prefix, key="fubrR".index))
```
`STRING_PREFIX_CHARS` could also be reordered to be used as the index string

Comment (felix-hilden) [reply src/black/strings.py:143]: That could actually be quite fun 😄 but with #2251 we are still second guessing the decisions, so this is a bit blocked.

Comment (jamesmyatt) [reply src/black/strings.py:143]: https://github.com/psf/black/pull/2740 has been merged now

File: src/black/strings.py
Comment (felix-hilden) [reply src/black/strings.py:154]: Also it short circuits the (likely) more complex subscript and `.lower`.

**Comment by jamesmyatt** [comment]
FWIW, I liked the `new_prefix = "".join(sorted(new_prefix, key="fbrR".index))` version, especially since "fbrR" could be replaced with a module-level variable, e.g. `PREFIX_ORDER`, which can then be modified externally if someone wants to hack it to change the order. OTOH, having options for changing the behaviour is not in black's MO.

**Comment by felix-hilden** [comment]
@jamesmyatt Fair enough, but I don't much care for how it reads. It's easy enough to read the current version as "if "r" isn't first, switch the characters around". Using a join and sorting by an index without a condition is quite a bit more complex and less efficient. And yeah, we won't introduce a hackable value 😄 although I don't know what'll stop a dedicated monkey patcher.

File: docs/the_black_code_style/current_style.md
Comment (JelleZijlstra) [inline docs/the_black_code_style/current_style.md:239]: This should be changed (somewhat unrelated to this PR): we only support formatting 3.6+, so we currently always remove the `u`.

**Review by JelleZijlstra — COMMENTED** [review]
I like r-first, since we say "raw f-string" not "f-raw string".

File: docs/the_black_code_style/current_style.md
Comment (felix-hilden) [reply docs/the_black_code_style/current_style.md:239]: Good catch 👍 We still support 3.3-3.5 right? Not that it changes the message but anyways.

Comment (JelleZijlstra) [reply docs/the_black_code_style/current_style.md:239]: Oh right, we do. But I'm not sure what the "3.6+" is about in the first place in that case, since u prefixes are redundant in all of Python 3. (And in Python 3.0-3.2 they did nothing.)

**Comment by felix-hilden** [comment]
The primer fail seems unrelated (thanks Jelle! ❤️) So this could now finally be ready for the last comments and a merge!

**Review by JelleZijlstra — APPROVED** [review]

**Comment by felix-hilden** [comment]
Last force to fix the primer after #2769

File: docs/the_black_code_style/current_style.md
Comment (cooperlees) [inline docs/the_black_code_style/current_style.md:239]: Love this and agree with all!

**Review by cooperlees — APPROVED** [review]
All looks good to me. I'm looking forward to this normalization. Thanks!

**Comment by ichard26** [comment]
Yay we did it, it only took half a year haha :)

**Comment by cooperlees** [comment]
> Yay we did it, it only took half a year haha :)

We can blame Python 2 ... Which we will never have to every again :D

**Comment by jamesmyatt** [comment]
Thanks so much @felix-hilden and @JelleZijlstra

**Pull request merged**
